### ISSUE TITLE
Support for props extraction on .vue

### ISSUE THREAD

**Issue by pikax** [johnsoncodehk/volar#94]
When you want to infer props from a base component, volar currently can't infer the `props` types (`vuedx` currently works) 

This code works in typescript
```ts
const TestBase = defineComponent({
  props: {
    show: Boolean,
    title: String,
  },
});

type TestBaseType = typeof TestBase extends DefineComponent<
  infer Props,
  any,
  any,
  any,
  any,
  any,
  any,
  any
>
  ? Props
  : {};

const Child = defineComponent({
  props: {
    // props contains the props options
    ...(TestBase.props as TestBaseType),

    extra: String,
  },
  setup(props) {
    let i: boolean = props.show; // boolean
  },
});
```

But as soon as you separate into SFC types, the type gets lost: 

```html
// test.vue
<script lang="ts">
import { defineComponent } from 'vue';

export default defineComponent({
  props: {
    show: Boolean,
    title: String,
  },
});
</script>

// Child.vue
<script lang="ts">
import { defineComponent } from 'vue';
import TestBase from './test.vue'

type TestBaseType = typeof TestBase extends DefineComponent<
  infer Props,
  any,
  any,
  any,
  any,
  any,
  any,
  any
>
  ? Props
  : {};

export default defineComponent({
  props: {
    // props contains the props options
    ...(TestBase.props as TestBaseType),

    extra: String,
  },
  setup(props) {
    // does not work beacuse TestBaseType will be `{}`
    let i: boolean = props.show; // boolean
  },
});
</script>
```

**Comment by johnsoncodehk** [comment]
This is the design limit for now, because `TestBase` types is:

```ts
DefineComponent<...> & {
  __VLS_options: ..., // for template components, props, emits services
  __VLS_slots: ..., // for slots services
}
```

So it can't infer extends as `DefineComponent`. A simple reproduce:

```ts
const a = defineComponent({
  props: {
    show: Boolean,
    title: String,
  },
});
type A = typeof a;
type B = A & { __VLS_foo: any }
// We want C is true, but it's false
type C = B extends DefineComponent<
  infer _,
  any,
  any,
  any,
  any,
  any,
  any,
  any
> ? true : false;
```

If you accept a hack way, you can use `typeof TestBase['__VLS_raw']` for now.

**Comment by pikax** [comment]
Will use that for now, do you think if the typed slots are implemented + GlobalComponents, will you still have the need of having?
```ts
DefineComponent<...> & {
  __VLS_raw: DefineComponent<...>, // raw type
  __VLS_options: ..., // for template components, props, emits services
  __VLS_slots: ..., // for slots services
}
```

**Comment by johnsoncodehk** [comment]
@pikax Will be improved, but still need to use for support vue 2 / class component.

I will find a way to handle this problem. (The most extreme case I can two typescript language service instances, one for `<script>`, one for `<template>`. :P)

**Comment by lanyizi** [comment]
I'm not sure if this bug has already been fixed, but the following one seems to be working:
```ts
import { ComponentOptions } from 'vue'
import Sfc from 'Sfc.vue'

type ExtractProp<Type> = Type extends ComponentOptions<infer Props>
  ? Props
  : never
type SfcProps = ExtractProp<typeof Sfc>;
```

**Comment by johnsoncodehk** [comment]
@lanyizi not yet, but this problem may only occur with `extends DefineComponent`.

**Comment by johnsoncodehk** [comment]
Fixed in volar@0.26.0 and vue-tsc@0.2.1.

**Comment by franco-onevillas** [comment]
Hey there, is it possible nowadays to infer types from a `.vue` file? I'm stuck in this exact case. Thanks!
